File: C/MaxSubarraySum/MaxSubarraySum.cpp
#include "MaxSubarraySum.h"


// Print the input array in a beautiful format
void print_array(int arr[], int N)
{
	int i;
	if (N <= OUTPUT_CUTOFF) {
		for (i = 0; i < N - 1; i++)
			printf("%6d, ", arr[i]);
	}
	else {
		for (i = 0; i < OUTPUT_CUTOFF / 2; i++)
			printf("%6d, ", arr[i]);
		printf("......, ");
		for (i = N - OUTPUT_CUTOFF / 2; i < N - 1; i++)
			printf("%6d, ", arr[i]);
	}
	printf("%6d\n", arr[N - 1]);
}


// Calculate the time between `end` and `start` in millisecond
double timing(clock_t end, clock_t start)
{
	return (double)(end - start) * 1000 / CLOCKS_PER_SEC;
}


// Solution for the Maximum Subarray Sum problem
// Exhaustively tries all possibilities for the solution to the problem
ElementType BruteForce(const ElementType arr[], int N)
{
	int i, j;
	ElementType temp, max_sum;

	max_sum = arr[0];
	for (i = 0; i < N; i++) {
		temp = 0;
		for (j = i; j < N; j++) {
			temp += arr[j];
			if (temp > max_sum) {
				max_sum = temp;
			}
		}
	}
	return max_sum;
}


static ElementType max3(ElementType a, ElementType b, ElementType c)
{
	ElementType temp;
	temp = (a > b) ? a : b;
	return (c > temp) ? c : temp;
}


static ElementType Recursive_Helper(const ElementType arr[], int left, int right)
{
	int center, i;
	ElementType temp;
	ElementType max_sum_left, max_sum_right;
	ElementType max_sum_left_border, max_sum_right_border;

	if (left == right)
		return arr[left];

	center = (left + right) / 2;
	max_sum_left = Recursive_Helper(arr, left, center);
	max_sum_right = Recursive_Helper(arr, center+1, right);

	temp = 0;
	max_sum_left_border = arr[center];
	for (i = center; i >= left; i--) {
		temp += arr[i];
		if (temp > max_sum_left_border) {
			max_sum_left_border = temp;
		}
	}

	temp = 0;
	max_sum_right_border = arr[center + 1];
	for (i = center+1; i <= right; i++) {
		temp += arr[i];
		if (temp > max_sum_right_border) {
			max_sum_right_border = temp;
		}
	}
	temp = max_sum_left_border + max_sum_right_border;
	return max3(max_sum_left, temp, max_sum_right);
}


// Solution for the Maximum Subarray Sum problem
// This algorithm uses a "Divide-and-Conquer" strategy
ElementType Recursive(const ElementType arr[], int N)
{
	return Recursive_Helper(arr, 0, N - 1);
}


// Solution for the Maximum Subarray Sum problem
// Implementation of Kadane's algorithm
ElementType Kadane(const ElementType arr[], int N)
{
	int i;
	ElementType temp, max_sum;

	temp = 0;
	max_sum = arr[0];
	for (i = 0; i < N; i++) {
		temp += arr[i];
		if (temp > max_sum) {
			max_sum = temp;
		}
		if (temp < 0) {
			temp = 0;
		}
	}
	return max_sum;
}

File: Python/MaxSubarraySum.py
"""
Provide solutions for the Maximum Subarray Sum problem
"""


__all__ = [
    "BruteForce",
    "Recursive",
    "Kadane",
]


def BruteForce(arr):
    """
    Solution for the Maximum Subarray Sum problem

    Exhaustively tries all possibilities for the solution to the problem

    Parameters
    ----------
    arr : array-like
        A one dimensional array of numbers.

    Returns
    -------
    max_sum : number
        The maximum subarray sum.

    Examples
    --------
    >>> from MaxSubarraySum import BruteForce
    >>> BruteForce([-2, 11, -4, 13, -5, -2])
    20
    >>> BruteForce([-5, -6, -7, -8, -9, -3, -6, -7])
    -3
    """

    length = len(arr)
    max_sum = arr[0]
    for i in range(length):
        temp = 0
        for j in range(i, length):
            temp += arr[j]
            if temp > max_sum:
                max_sum = temp
    return max_sum


def _Recursive_Helper(arr, left, right):
    if left == right:
        return arr[left]

    center = (left + right) // 2
    left_max = _Recursive_Helper(arr, left, center)
    right_max = _Recursive_Helper(arr, center + 1, right)
    temp = 0
    left_border_max = arr[center]
    for i in range(center, left-1, -1):
        temp += arr[i]
        if temp > left_border_max:
            left_border_max = temp
    temp = 0
    right_border_max = arr[center+1]
    for i in range(center+1, right+1, 1):
        temp += arr[i]
        if temp > right_border_max:
            right_border_max = temp
    return max(left_max, left_border_max + right_border_max, right_max)


def Recursive(arr):
    """
    Solution for the Maximum Subarray Sum problem

    This algorithm uses a "Divide-and-Conquer" strategy.

    Parameters
    ----------
    arr : array-like
        A one dimensional array of numbers.

    Returns
    -------
    max_sum : number
        The maximum subarray sum.

    Examples
    --------
    >>> from MaxSubarraySum import Recursive
    >>> Recursive([-2, 11, -4, 13, -5, -2])
    20
    >>> Recursive([-5, -6, -7, -8, -9, -3, -6, -7])
    -3
    """

    return _Recursive_Helper(arr, 0, len(arr) - 1)


def Kadane(arr):
    """
    Solution for the Maximum Subarray Sum problem

    Implementation of Kadane's algorithm

    Parameters
    ----------
    arr : array-like
        A one dimensional array of numbers.

    Returns
    -------
    max_sum : number
        The maximum subarray sum.

    Examples
    --------
    >>> from MaxSubarraySum import Kadane
    >>> Kadane([-2, 11, -4, 13, -5, -2])
    20
    >>> Kadane([-5, -6, -7, -8, -9, -3, -6, -7])
    -3
    """

    temp = 0
    max_sum = arr[0]
    for item in arr:
        temp += item
        if temp > max_sum:
            max_sum = temp
        if temp < 0:
            temp = 0
    return max_sum


# This is a test of this module
if __name__ == "__main__":
    import numpy as np

    from time import time

    size = 1000
    low = -size//2
    high = size//2
    array = np.random.randint(low, high, size)
    # array = [-2, 11, -4, 13, -5, -2]
    # array = (-5, -6, -7, -8, -9, -3, -6, -7)

    algorithms = {
        "Kadane": Kadane,
        "Recursive": Recursive,
        "Brute Force": BruteForce,
    }

    info0 = "The time spend on the {0} algorithm: {1:.3f}ms"
    info1 = "The result of the {0} algorithm: {1}"
    separator = "=" * max(len(info0), len(info1)) + '\n'

    buff = []
    for algorithm, func in algorithms.items():
        start = time()
        res = func(array)
        end = time()
        buff.append(res)
        print(info0.format(algorithm, 1000*(end-start)))
        print(info1.format(algorithm, res))
        print(separator)

    # Assert all the algorithms give the same answer
    for item in buff:
        assert item == buff[0]
